### PR TITLE
Prevent to delete to "bundler-" prefix gem like bundler-audit

### LINKED ISSUE #2081: gem update --system (2.7.2) deletes bundler-audit and bundler-patch gems
I'm having a problem or would like to suggest a feature.

My current problem is that `bundler-audit` and `bundler-patch` gems are deleted when I run `gem update --system`.

I can reproduce this bug on a brand new system using rbenv. I start with a fresh install of Ruby 2.4.2.

```
$ ruby -v
ruby 2.4.2p198 (2017-09-14 revision 59899) [x86_64-darwin17]
$ gem env version
2.6.13
```

Then I install the `bundler-audit` gem. Note that the gem and its binstub exist as expected.

```
$ gem install bundler-audit
Fetching: thor-0.20.0.gem (100%)
Successfully installed thor-0.20.0
Fetching: bundler-1.16.0.gem (100%)
Successfully installed bundler-1.16.0
Fetching: bundler-audit-0.6.0.gem (100%)
Successfully installed bundler-audit-0.6.0
3 gems installed

$ gem list

*** LOCAL GEMS ***

bigdecimal (default: 1.3.0)
bundler (1.16.0)
bundler-audit (0.6.0)
did_you_mean (1.1.0)
io-console (default: 0.4.6)
json (default: 2.0.4)
minitest (5.10.1)
net-telnet (0.1.1)
openssl (default: 2.0.5)
power_assert (0.4.1)
psych (default: 2.2.2)
rake (12.0.0)
rdoc (default: 5.0.0)
test-unit (3.2.3)
thor (0.20.0)
xmlrpc (0.2.1)

$ ls -ld ~/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/bundler-audit-0.6.0/bin/bundle-audit 
.rwxrwxr-x 228 mattb 17 Nov  9:55 /Users/mattb/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/bundler-audit-0.6.0/bin/bundle-audit
```

Now I upgrade to rubygems 2.7.2. Notice the warnings that are printed.

```
$ gem update --system
Updating rubygems-update
Fetching: rubygems-update-2.7.2.gem (100%)
Successfully installed rubygems-update-2.7.2
Installing RubyGems 2.7.2
Bundler 1.16.0 installed
RubyGems 2.7.2 installed
Regenerating binstubs
`/Users/mattb/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/bundler-audit-0.6.0/bin/bundle-audit` does not exist, maybe `gem pristine bundler-audit` will fix it?
`/Users/mattb/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/bundler-audit-0.6.0/bin/bundler-audit` does not exist, maybe `gem pristine bundler-audit` will fix it?
RubyGems system software updated
```

The `bundler-audit` gem has been deleted!

```
$ ls -ld ~/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/bundler-audit-0.6.0/bin/bundle-audit 
"/Users/mattb/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/bundler-audit-0.6.0/bin/bundle-audit": No such file or directory (os error 2)

$ ls -l ~/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/
drwxr-xr-x - mattb 17 Nov  9:56 bundler-1.16.0
drwxr-xr-x - mattb 17 Nov  9:54 did_you_mean-1.1.0
drwxr-xr-x - mattb 17 Nov  9:54 minitest-5.10.1
drwxr-xr-x - mattb 17 Nov  9:54 net-telnet-0.1.1
drwxr-xr-x - mattb 17 Nov  9:54 power_assert-0.4.1
drwxr-xr-x - mattb 17 Nov  9:54 rake-12.0.0
drwxr-xr-x - mattb 17 Nov  9:54 rdoc-5.0.0
drwxr-xr-x - mattb 17 Nov  9:56 rubygems-update-2.7.2
drwxr-xr-x - mattb 17 Nov  9:54 test-unit-3.2.3
drwxr-xr-x - mattb 17 Nov  9:55 thor-0.20.0
drwxr-xr-x - mattb 17 Nov  9:54 xmlrpc-0.2.1
```

But strangely it still appears in the gem list:

```
$ gem list

*** LOCAL GEMS ***

bigdecimal (default: 1.3.0)
bundler (default: 1.16.0)
bundler-audit (0.6.0)
did_you_mean (1.1.0)
io-console (default: 0.4.6)
json (default: 2.0.4)
minitest (5.10.1)
net-telnet (0.1.1)
openssl (default: 2.0.5)
power_assert (0.4.1)
psych (default: 2.2.2)
rake (12.0.0)
rdoc (default: 5.0.0)
rubygems-update (2.7.2)
test-unit (3.2.3)
thor (0.20.0)
xmlrpc (0.2.1)
```

In my testing this happens with `bundler-patch` as well. I'm guessing it has something to do with the `bundler-` prefix. Other gems don't seem to be affected.

---

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [ ] Publishing a library
  - [x] The command line `gem`
  - [ ] Other

Here are my current environment details:

```
$ gem env version
2.7.2
```

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### PR DESCRIPTION
**Pull request by hsbt**
Fixes #2081.

### CODE REVIEW

File: test/rubygems/test_gem_commands_setup_command.rb [hunk 1]
```diff
@@ -27,6 +27,28 @@ def setup
     open 'bundler/exe/bundle',        'w' do |io| io.puts '# bundle'       end
     open 'bundler/lib/bundler.rb',    'w' do |io| io.puts '# bundler.rb'   end
     open 'bundler/lib/bundler/b.rb',  'w' do |io| io.puts '# b.rb'         end
+
+    FileUtils.mkdir_p 'default/gems'
+
+    gemspec = Gem::Specification.new
+    gemspec.name = "bundler"
```
Comment (lucasprag) [inline test/rubygems/test_gem_commands_setup_command.rb]: Prefer single-quoted strings when you don't need string interpolation of special symbols.

File: lib/rubygems/commands/setup_command.rb [hunk 2]
```diff
@@ -373,7 +373,7 @@ def install_default_bundler_gem
     bundler_spec = Gem::Specification.load(default_spec_path)
 
     Dir.entries(bundler_spec.gems_dir).
-      select {|default_gem| default_gem.start_with?("bundler-") }.
+      select {|default_gem| default_gem.match(/bundler-\d/) }.
```
Comment (segiddins) [inline lib/rubygems/commands/setup_command.rb]: maybe `default_gem.match(/^bundler-#{Gem::Version::VERSION_PATTERN}$/)`?

**Comment by hsbt** [comment]
@bundlerbot r+

**Comment by hsbt** [comment]
@bundlerbot retry

**Comment by hsbt** [comment]
@bundlerbot r+

File: lib/rubygems/commands/setup_command.rb (hunk 2 above)
Comment (segiddins) [reply lib/rubygems/commands/setup_command.rb]: you didn't include the `^`, so `foo-bundler-1` could still be deleted

Comment (hsbt) [reply lib/rubygems/commands/setup_command.rb]: `default_gem` returns `/full/path/of/gems/bundler-x.y.z` . I think it needs to `File.basename(default_gem).match(/.../)` .

I will revise it with changes of other PRs.

**Pull request merged**